go/libraries/doltcore/env/dolt_docs.go

```diff
@@ -100,3 +103,145 @@ func hasDocFile(fs filesys.ReadWriteFS, file string) bool {
+// repo yet.
+func UpdateRootWithDocsTable(ctx context.Context, dbData DbData, root *doltdb.RootValue, rootType RootType, docDetails []doltdb.DocDetails) (*doltdb.RootValue, error) {
+	docTbl, _, err := root.GetTable(ctx, doltdb.DocTableName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	docTbl, err = dbData.Drw.WriteDocsToDisk(ctx, root.VRW(), docTbl, docDetails)
+
+	if errors.Is(ErrEmptyDocsTable, err) {
+		root, err = root.RemoveTables(ctx, doltdb.DocTableName)
+	} else if err != nil {
+		return nil, err
+	}
+
+	// There might not need be a need to create docs table if not docs have been created yet so check if docTbl != nil.
+	if docTbl != nil {
+		root, err = root.PutTable(ctx, doltdb.DocTableName, docTbl)
+	}
+
+	switch rootType {
+	case Working:
+		_, err = UpdateWorkingRoot(ctx, dbData.Ddb, dbData.Rsw, root)
+	case Staged:
+		_, err = UpdateStagedRoot(ctx, dbData.Ddb, dbData.Rsw, root)
+	default:
+		return nil, errors.New("Root type not supported with docs update.")
+	}
+	return root, nil
+}
+
+// ResetWorkingDocsToStagedDocs resets the `dolt_docs` table on the working root to match the staged root.
+// If the `dolt_docs` table does not exist on the staged root, it will be removed from the working root.
+func ResetWorkingDocsToStagedDocs(ctx context.Context, ddb *doltdb.DoltDB, rsr RepoStateReader, rsw RepoStateWriter) error {
+	wrkRoot, err := WorkingRoot(ctx, ddb, rsr)
+	if err != nil {
+		return err
+	}
+
+	stgRoot, err := StagedRoot(ctx, ddb, rsr)
+	if err != nil {
+		return err
+	}
+
+	stgDocTbl, stgDocsFound, err := stgRoot.GetTable(ctx, doltdb.DocTableName)
+	if err != nil {
+		return err
+	}
+
+	_, wrkDocsFound, err := wrkRoot.GetTable(ctx, doltdb.DocTableName)
+	if err != nil {
+		return err
+	}
+
+	if wrkDocsFound && !stgDocsFound {
+		newWrkRoot, err := wrkRoot.RemoveTables(ctx, doltdb.DocTableName)
+		if err != nil {
+			return err
+		}
+		_, err = UpdateWorkingRoot(ctx, ddb, rsw, newWrkRoot)
+		return err
+	}
+
+	if stgDocsFound {
+		newWrkRoot, err := wrkRoot.PutTable(ctx, doltdb.DocTableName, stgDocTbl)
+		if err != nil {
+			return err
+		}
+		_, err = UpdateWorkingRoot(ctx, ddb, rsw, newWrkRoot)
+		return err
+	}
+	return nil
+}
+
+// GetDocsWithNewerTextFromRoot returns Docs with the NewerText value(s) from the provided root. If docs are provided,
+// only those docs will be retrieved and returned. Otherwise, all valid doc details are returned with the updated NewerText.
+func GetDocsWithNewerTextFromRoot(ctx context.Context, root *doltdb.RootValue, docs Docs) (Docs, error) {
+	docTbl, docTblFound, err := root.GetTable(ctx, doltdb.DocTableName)
+	if err != nil {
+		return nil, err
+	}
+
+	var sch schema.Schema
+	if docTblFound {
+		docSch, err := docTbl.GetSchema(ctx)
+		if err != nil {
+			return nil, err
+		}
+		sch = docSch
+	}
+
+	if docs == nil {
+		docs = *AllValidDocDetails
+	}
+
+	for i, doc := range docs {
+		doc, err = doltdb.AddNewerTextToDocFromTbl(ctx, docTbl, &sch, doc)
```

Comment:
This should not be a method on doltdb

Reply:
NewerText is a bad name. Get rid of NewerText from the DocDetails object all together. If we want to compare to versions of a doc, we should use two different DocDetails objects.

Reply:
Move DocDetails out of the doltdb package in general.

Reply:
None of the methods that manipulate docs in general should be in the root_val.go file